src/OpenSpaceToolkit/Physics/Environment/Atmospheric/Earth/NRLMSISE00.cpp

```diff
@@ -0,0 +1,291 @@
+
+    const Time timeOfDay = currentDateTime.getTime();
+    const Integer secondsInDay = timeOfDay.getHour() * 3600 + timeOfDay.getMinute() * 60 + timeOfDay.getSecond();
+
+    // Solar flux values
+    const Real f107Previous = spaceWeatherManager.getF107SolarFluxAt(anInstant - Duration::Days(1));
+    const Real f107Average = spaceWeatherManager.getF107SolarFlux81DayAvgAt(anInstant);
+
+    // Ap solar indices array
+    this->computeApArray(aph.a, anInstant);
+
+    Real lst = Real::Undefined();
+
+    // Use actual sun position to compute local solar time if provided
+    if (aSunPosition.isDefined())
+    {
+        const Position positionITRF = {
+            aLLA.toCartesian(
+                EarthGravitationalModel::WGS84.equatorialRadius_, EarthGravitationalModel::WGS84.flattening_
+            ),
+            Position::Unit::Meter,
+            Frame::ITRF()};
+
+        const Position sunPositionITRF = aSunPosition.inFrame(Frame::ITRF(), anInstant);
+
+        lst = (Real::Pi() + std::atan2(
+                                sunPositionITRF.accessCoordinates()[0] * positionITRF.accessCoordinates()[1] -
+                                    sunPositionITRF.accessCoordinates()[1] * positionITRF.accessCoordinates()[0],
+                                sunPositionITRF.accessCoordinates()[0] * positionITRF.accessCoordinates()[0] +
+                                    sunPositionITRF.accessCoordinates()[1] * positionITRF.accessCoordinates()[1]
+                            )) *
+              12.0 / Real::Pi();
+    }
+    else
+    {
+        // This is the preferred method per the NRLMSISE documentation
+        // https://github.com/magnific0/nrlmsise-00/blob/master/nrlmsise-00.h#L103
+        lst = Real::Integer(secondsInDay) / 3600.0 + aLLA.getLongitude().inDegrees() / 15.0;
+    }
+
+    input.doy = dayOfYear;
+    input.year = year;
+    input.sec = secondsInDay;
+    input.alt = aLLA.getAltitude().inKilometers();
+    input.g_lat = aLLA.getLatitude().inDegrees();
+    input.g_long = aLLA.getLongitude().inDegrees();
+    input.lst = lst;
+    input.f107A = f107Average;
+    input.f107 = f107Previous;
+    input.ap = aph.a[0];
+    input.ap_a = &aph;
+}
+
+Real NRLMSISE00::getDensityAt(const LLA& aLLA, const Instant& anInstant, const Position& aSunPosition) const
+{
+    // Included from NRLMSISE-00.h
+    NRLMSISE00_c::ap_array ap_values_c;
+    NRLMSISE00_c::nrlmsise_input input_c;
+
+    NRLMSISE00_c::nrlmsise_output output;
+    NRLMSISE00_c::nrlmsise_flags flags;
+
+    // Redefined from NRLMSISE00.hpp
+    NRLMSISE00::nrlmsise_input input;
+    NRLMSISE00::ap_array ap_values;
+
+    // Set model behavior flags. Their meaning is defined in the NRLMSISE-00 header file.
+    // https://github.com/magnific0/nrlmsise-00/blob/master/nrlmsise-00.h
+    for (int i = 0; i < 24; i++)
+    {
+        flags.switches[i] = 1;
+    }
+
+    this->computeNRLMSISE00Input(input, ap_values, aLLA, anInstant, aSunPosition);
```

Comment:
Maybe this function can return the `input` and take `ap_values` as a strict input. Then we can add `computeApArray` above this as a separate call, so as to separate concerns in the functions.